### PR TITLE
Test for CVE-2016-7054: ChaCha20/Poly1305 heap-buffer-overflow

### PR DESCRIPTION
**Pull request by mildass**
Based on [PR no. 94](https://github.com/tomato42/tlsfuzzer/pull/94) created by @silverfoxy

### CODE REVIEW

File: scripts/test-cve-2016-7054.py [hunk 1]
```diff
@@ -0,0 +1,151 @@
...
+    node = node.add_child(ExpectChangeCipherSpec())
+    node = node.add_child(ExpectFinished())
+    node = node.add_child(ApplicationDataGenerator(
+            b"GET / HTTP/1.0\r\n" + b'X-test: ' + n * b'A'))
+    node = node.add_child(ExpectApplicationData())
+    node = node.add_child(AlertGenerator(AlertLevel.warning,
+                                         AlertDescription.close_notify))
+    node = node.add_child(ExpectAlert())
+    node.next_sibling = ExpectClose()
+    conversations["Correct 16-byte tag"] = \
```
Comment (tomato42) [inline scripts/test-cve-2016-7054.py]: Isn't this basically a "sanity" test case?

File: scripts/test-cve-2016-7054.py [hunk 2]
```diff
@@ -0,0 +1,151 @@
...
+    else:
+        run_only = None
+
+    conversations = {}
+
+    # 16 chars: POLY1305 tag 128 bit
+    # Tampering one bit suffices to damage the mac
+    # The payload has to be long enough to trigger heap overflow
+    n = 15000
+    fuzzes = [(x, 1) for x in range(-17, -1)]
```
Comment (tomato42) [inline scripts/test-cve-2016-7054.py]: but here you're modifying just he least significant bit of every byte, not all the bits of the tag

File: scripts/test-cve-2016-7054.py [hunk 3]
```diff
@@ -0,0 +1,151 @@
...
+        run_only = set(args)
+    else:
+        run_only = None
+
+    conversations = {}
+
+    # 16 chars: POLY1305 tag 128 bit
+    # Tampering one bit suffices to damage the mac
+    # The payload has to be long enough to trigger heap overflow
+    n = 15000
```
Comment (tomato42) [inline scripts/test-cve-2016-7054.py]: where did that number come from?

File: scripts/test-cve-2016-7054.py [hunk 4]
```diff
@@ -0,0 +1,151 @@
...
+
+    # run the conversation
+    good = 0
+    bad = 0
+    failed = []
+
+    for conversation_name in conversations:
+        conversation = conversations[conversation_name]
+        print(conversation_name + "...")
+        runner = Runner(conversation)
```
Comment (tomato42) [inline scripts/test-cve-2016-7054.py]: special handling of the "sanity" test case is missing

File: scripts/test-cve-2016-7054.py (hunk 3 above)
Comment (mildass) [inline scripts/test-cve-2016-7054.py]: At CVE specification is:
> corrupting larger payloads
So I used same length as was length used at old PR. What length should I use? `2**14 - 24` as max. length - 24 bytes (GET request length + "X-test: ")?

Comment (tomato42) [inline scripts/test-cve-2016-7054.py]: mostly, explain where it came from. I don't know what length is necessary to test

File: scripts/test-cve-2016-7054.py [hunk 5]
```diff
@@ -0,0 +1,163 @@
...
+    node = node.add_child(ExpectCertificate())
+    node = node.add_child(ExpectServerKeyExchange())
+    node = node.add_child(ExpectServerHelloDone())
+    node = node.add_child(ClientKeyExchangeGenerator())
+    node = node.add_child(ChangeCipherSpecGenerator())
+    node = node.add_child(FinishedGenerator())
+    node = node.add_child(ExpectChangeCipherSpec())
+    node = node.add_child(ExpectFinished())
+    node = node.add_child(ApplicationDataGenerator(
+            b"GET / HTTP/1.0\r\n" + b'X-test: ' + n * b'A'))
```
Comment (tomato42) [inline scripts/test-cve-2016-7054.py]: the `\r\n\r\n` is missing at the end to make it a valid HTTP request (that's why the test against tlslite-ng fails, it has a proper HTTP server, so won't answer to requests until they are finished)

also add it to the error test case, so they are as similar as possible

**Comment by tomato42** [comment]
Thanks! Sorry the second review and merge took me so long!

**Pull request merged**